### DISCUSSION TITLE
RegExp objects in allowedPostUpgradeCommands in config.js get replaced with empty objects

### DISCUSSION THREAD

**Discussion by jpgoelz** [renovatebot/renovate#25792] [Request Help]
### How are you running Renovate?

Self-hosted Renovate

### If you're self-hosting Renovate, tell us what version of Renovate you run.

37.59.0

### If you're self-hosting Renovate, select which platform you are using.

GitLab self-hosted

### Was this something which used to work for you, and then stopped?

I am trying to get this working for the first time

### Describe the problem

We were trying to use RegExp objects in the array for `allowedPostUpgradeCommands` in our config.js.

We received an error as an issue, that the regex was not valid:

![image](https://github.com/renovatebot/renovate/assets/32487587/e01c4c1d-a8d3-49a5-80cb-eced89e8ff8b)

What struck me as odd was the fact that the regex was printed as `[object Object]` when RegExp#toString() should normally print the regex string.

I did some digging (several hours) and found out, that the `migrateConfig()` function in `lib/config/migration.ts` replaces the RegExp objects with empty object.

I logged the `config` object and the `migratedConfig` objects to the console:

`config`:
![image](https://github.com/renovatebot/renovate/assets/32487587/3bfb8973-bff5-4893-924b-1f480a99b2e4)

`migratedConfig`:
![image](https://github.com/renovatebot/renovate/assets/32487587/fcb2eb2b-d0f5-4b91-91da-ebe1bc092a96)

It turns out that because `allowedPostUpgradeCommands` is an array, we land at this portion in the code:
https://github.com/renovatebot/renovate/blob/d45f437f2a58323a05fbf9824454029296c7231d/lib/config/migration.ts#L71-L82

And because `item` is our RegExp object (e.g. `/my regex/`) the method `migrateConfig()` is called with the item as a config, in which case we land at

https://github.com/renovatebot/renovate/blob/d45f437f2a58323a05fbf9824454029296c7231d/lib/config/migration.ts#L83-L88

Which ultimately replaces the valid RegExp object with an empty `{}`, when the following line is executed:
https://github.com/renovatebot/renovate/blob/d45f437f2a58323a05fbf9824454029296c7231d/lib/config/migration.ts#L32
Debugger in `lib/config/migrations/migrations-service.ts`:
![image](https://github.com/renovatebot/renovate/assets/32487587/290e9095-7d89-44f1-9b9d-8e183371f21c)

Is there a good reason behind this?

### Have you created a minimal reproduction repository?

Placeholder value, please select the correct response from the dropdown

**Comment by jpgoelz** [comment]
I have already found a solution and can provide a pull request if desired.

**Reply by jpgoelz** [reply]
I went ahead and created the PR: https://github.com/renovatebot/renovate/pull/25793.

**Reply by rarkins** [reply]
@jpgoelz let's discuss here instead

**Comment by rarkins** [comment]
Renovate does not support regex objects, the correct solution is to raise a config error if this is detected

**Reply by jpgoelz** [reply]
Hi @rarkins, thank you for taking the time!

> Renovate does not support regex objects, the correct solution is to raise a config error if this is detected

Could you elaborate on the reasoning behind that? Being able to use RegExp directly would give us the benefit of validation by the IDE.

It took us several hours yesterday trying to find the problem. 
The RegExp is replaced by `{}` and so the config error is only raised in the `regEx()` function in `/lib/util/regex.ts` where `pattern` is expected to be `string | RegExp` but not `{}` which is then printed using the `toString()` method resulting in `[object Object]`. When developing in that method, I would not expect an empty object to be passed.

Also, I think it would be best to raise the error when reading the config file on startup (fail fast!). Raising it that late is why we did not find it quickly, because we were expecting it in our repository's configuration. The issue created by Renovate did not tell us that the origin of the error lay in the general Renovate config (location was also `[object Objecct]`, see original post).
The error only came up when our package rule using `postUpgradeTasks` was applied. So of course we tried to find the mistake there at first.

Just a quick idea: How about not replacing "illegal" objects with empty objects, but rather with an Error object that holds details about what is wrong. For this a custom migration could be added. This Error could then be read and printed as a more meaningful message in the issue. However, I still believe the "wrong" configuration should have Renovate fail on startup, not only if a matching package has an upgrade.

**Comment by rarkins** [comment]
> > Renovate does not support regex objects, the correct solution is to raise a config error if this is detected
> 
> Could you elaborate on the reasoning behind that? Being able to use RegExp directly would give us the benefit of validation by the IDE.

For simplicity of implementation and consistency across configuration approaches.
 
> Also, I think it would be best to raise the error when reading the config file on startup (fail fast!). 

Yes, that's what I suggested

> Just a quick idea: How about not replacing "illegal" objects with empty objects, but rather with an Error object that holds details about what is wrong. 

I prefer to just "fail fast" as you put it.

**Reply by jpgoelz** [reply]
> Yes, that's what I suggested

Ah, I thought with 

> [...] raise a config error if this is detected

you meant "when the regex is parsed when applying the commands", which would be late.

**Reply by rarkins** [reply]
No, I think we should validate and fail early. Today we are too loose/permissive with global config parsing

**Reply by jpgoelz** [reply]
Is there any way I can help?

**Reply by rarkins** [reply]
You're welcome to submit a PR which detects a regex object or other invalid objects and errors out. We should merge this in v38 though as it will technically be breaking
